File: week13/7/[241129](Leet1026)_Maximum Difference Between Node and Ancestor.cpp
/*"""Given the root of a binary tree, find the maximum value v for which there exist different nodes a and b where v = |a.val - b.val| and a is an ancestor of b.

A node a is an ancestor of b if either: any child of a is equal to b or any child of a is an ancestor of b."""

"""Constraints:
    The number of nodes in the tree is in the range [2, 5000].
    0 <= Node.val <= 105"""*/

/*수도코드1 : 모든 노드에 대해서 비교하는 과정에서 BFS 활용
목표 : 각 노드의 subtree에 있는 모든 노드들과의 max_diff 찾는 함수를 정의해서, 해당 함수를 Tree의 모든 노드에 대해 실행하면서 main_max_diff를 update
*/

#include <iostream>
#include <queue>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int cur_max_diff(TreeNode* cur_node){
        int max_diff = 0;

        queue<TreeNode*> q;
        q.push(cur_node);

        while (!q.empty()){
            TreeNode* cur = q.front();
            q.pop();

            int diff = cur_node->val - cur->val;
            int abs_diff = diff > 0 ? diff : -diff;
            if (abs_diff > max_diff) max_diff = abs_diff;

            if (cur->left != nullptr) q.push(cur->left);
            if (cur->right != nullptr) q.push(cur->right);
        }

        return max_diff;
    }

    int maxAncestorDiff(TreeNode* root) {
        int main_max_diff = 0;

        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()){
            TreeNode* cur = q.front();
            q.pop();

            int cur_max_d = cur_max_diff(cur);
            if (cur_max_d > main_max_diff) main_max_diff = cur_max_d;

            if (cur->left != nullptr) q.push(cur->left);
            if (cur->right != nullptr) q.push(cur->right);
        }

        return main_max_diff;
    }
};


/*DFS 활용 풀이---------------------------------------------------------------------------------------------------------*/

/*수도코드2
목표 : root에서부터 각 leaf까지의 경로를 따라 내려가면서 big, small값을 갱신하고, 경로의 끝에 도달하면 diffrence(big-small) 값을 반환하는 help함수를 정의함. root가 nullptr인 경우에는 0을 반환하고 그 외에는 helper함수를 사용해서 값을 얻는다.
*/

class Solution {
public:
    int helper(TreeNode* curNode, int min_val, int max_val){
        if (curNode == nullptr) return max_val - min_val;

        if (curNode->val > max_val) max_val = curNode->val;
        if (curNode->val < min_val) min_val = curNode->val;

        int left_max_diff = helper(curNode->left, min_val, max_val);
        int right_max_diff = helper(curNode->right, min_val, max_val);

        return left_max_diff > right_max_diff ? left_max_diff : right_max_diff;
    }

    int maxAncestorDiff(TreeNode* root) {
        if (root == nullptr) return 0;

        return helper(root, root->val, root->val);
    }
};

File: week13/7/[241129](Leet1026)_Maximum Difference Between Node and Ancestor.py
"""Given the root of a binary tree, find the maximum value v for which there exist different nodes a and b where v = |a.val - b.val| and a is an ancestor of b.

A node a is an ancestor of b if either: any child of a is equal to b or any child of a is an ancestor of b."""

"""Constraints:
    The number of nodes in the tree is in the range [2, 5000].
    0 <= Node.val <= 105"""
    
"""아이디어
    Helper함수 : 현재 Node(root)에서 모든 자식들과의 diff를 비교하고 max_diff값을 반환.하는 함수
        while 자녀가 있는 동안
            cur을 갱신하면서 
            abs(root-cur) 값을 계산.
            만약 abs(root-cur) 값이 max_diff보다 더 크면 max_diff를 갱신한다.

    BFS(root)에 이면 내 하위 Node를 다 볼 수 있으니까...
    BFS하면서 max_dif를 update하는 방식으로 해야겠다.
"""

from collections import deque

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution(object):
    def maxAncestorDiff(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        def cur_max_diff(cur_top):
            max_dif = 0

            q = deque([cur_top])
            while q :
                cur = q.popleft()
                
                if abs(cur_top.val - cur.val) > max_dif :
                    max_dif = abs(cur_top.val - cur.val)
                
                if cur.left :
                    q.append(cur.left)
                if cur.right :
                    q.append(cur.right)
            
            return max_dif
        
        max_difference= 0
        
        q = deque([root])
        while q:
            cur_node = q.popleft()
            cur_max = cur_max_diff(cur_node)
            if cur_max > max_difference :
                max_difference = cur_max
                
            if cur_node.left :
                q.append(cur_node.left)
            if cur_node.right :
                q.append(cur_node.right)
        
        return max_difference
        

# [다른 방법1] self.diff + DFS 활용 root->leaf 까지의 path 하나 당 (big, small) 추적-----------------------------------------------------------------------------------

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution(object):
    def maxAncestorDiff(self, root):
        if not root:
            return 0
        self.diff = 0 #트리 전체에서 발견된 최대 차이를 저장하는 변수
        self.helper(root, root.val, root.val)
        return self.diff
    
    def helper(self, root, min_val, max_val): #재귀적으로 트리 탐색하면서 각 노드에서의 최소값, 최대값 업데이트, 최대 차이 계산
        if not root: #현재 노드가 None이면 재귀호출 종료
            return
        # 현재노드의 값과 최소값 / 또는 현재노드의 값과 최대값 차이를 계산해서 self.diff update
        self.diff = max(self.diff, max(abs(min_val - root.val), abs(max_val - root.val)))
        #현재 노드값을 활용해서 min_val 또는 max_val 업데이트
        min_val = min(min_val, root.val)
        max_val = max(max_val, root.val)
        #왼쪽 자식과 오른쪽 자식에 대해 재귀적으로 helper함수 호출
        self.helper(root.left, min_val, max_val)
        self.helper(root.right, min_val, max_val)



# [다른방법 2] DFS활용-------------------------------------------------------

class Solution(object):
    def maxAncestorDiff(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self.helper(root, root.val, root.val)

    def helper(self, r, mn, mx): #return으로 max_diff값을 반환하는 형태.
        if not r:
            return mx - mn # 더이상 내려갈 자식이 없을 때 해당 경로에서의 max_diff 반환.
        mn = min(mn, r.val)
        mx = max(mx, r.val)

        left_diff = self.helper(r.left, mn, mx)
        right_diff = self.helper(r.right, mn, mx)

        return max(left_diff, right_diff)